File: LEETCODE_1462_Course_Schedule_IV.cpp
#include<bits/stdc++.h>
using namespace std;
class Solution {
    bool bfs(int& source, int& destination, unordered_map<int,bool>& visited, unordered_map<int,vector<int>>& graph) {
        queue<int> q;
        q.push(source);
        visited[source] = true;
        while(!q.empty()) {
            int node = q.front();
            if(node == destination) {
                return true;
            }
            q.pop();
            for(int& neighbour: graph[node]) {
                if(visited[neighbour] == false) {
                    q.push(neighbour);
                    visited[neighbour] = true;
                }
            }
        }
        return false;
    }
public:
    vector<bool> checkIfPrerequisite(int numCourses, vector<vector<int>>& prerequisites, vector<vector<int>>& queries) {
        unordered_map<int,vector<int>> graph;
        for(vector<int>& vec: prerequisites) {
            graph[vec[0]].push_back(vec[1]);
        }
        vector<bool> answer;
        for(vector<int>& query: queries) {
            unordered_map<int,bool> visited;
            answer.push_back(this->bfs(query[0], query[1], visited, graph));
        }
        return answer;
    }
};

File: LEETCODE_1462_Course_Schedule_IV.java
import java.util.*;

class LEETCODE_1462_Course_Schedule_IV {

    private boolean bfs(int source, int destination, Set<Integer> visited, Map<Integer, List<Integer>> graph) {
        ArrayDeque<Integer> q = new ArrayDeque<>();
        q.offerLast(source);
        visited.add(source);
        while(!q.isEmpty()) {
            int node = q.pollFirst();
            if(node == destination) {
                return true;
            }
            if(graph.containsKey(node)) {
                for(int neighbour: graph.get(node)) {
                    if(!visited.contains(neighbour)) {
                        q.offerLast(neighbour);
                        visited.add(neighbour);
                    }
                }
            }
        }
        return false;
    }

    public List<Boolean> checkIfPrerequisite(int numCourses, int[][] prerequisites, int[][] queries) {
        Map<Integer, List<Integer>> graph = new HashMap<>();
        for(int[] vec: prerequisites) {
            if(!graph.containsKey(vec[0])) {
                graph.put(vec[0], new ArrayList<>());
            }
            graph.get(vec[0]).add(vec[1]);
        }
        List<Boolean> answer = new ArrayList<>();
        for(int[] query: queries) {
            Set<Integer> visited = new HashSet<>();
            answer.add(this.bfs(query[0], query[1], visited, graph));
        }
        return answer;
    }
}